RFD 0123 TLS Routing behind load balancers (#22347)
gravitational/teleport: rfd/0xxx-tls-routing-behind-lb.md

                        │                 │
                    ┌───▼─────┐        ┌──▼──────┐
                    │Handshake│        │Handshake├───────────────┐
┌───────────────────┤ Success │        │ Failed  │               │
│                   └───┬─────┘        └┬────────┘               │
│                       │               │                        │
│Negotiated             │No ALPN        │remote error: tls:      │other
│teleport-reversetunnel │Negotiated     │no application protocol │errors
│                       │               │                        │
│                   ┌───▼────┐          │                  ┌─────▼──┐
│                   │Upgrade ◄──────────┘                  │  Not   │
│                   │Required│                             │Required│
│                   └────────┘                             └────▲───┘
│                                                               │
└───────────────────────────────────────────────────────────────┘
```
Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,
hinting there is a load balancer in the middle that terminates TLS.

> However, this test may not be bullet proof. Different load balancers (like the one million [Kubernetes Ingress

Author: added explanation on the remote message and added a `tls_routing_upgrade_mode` to webapi ping.

clusters to single port mode if the cluster is currently serving separate ports.

All scenarios discussed in this section assumes Teleport sits behind a load balancer so connection upgrade for TLS
Routing is required.

#### 1 - Database Access UX

Common flows already use `tsh` local proxy (since v8) so they will automatically perform the connection upgrade:

- `tsh db connect` (single port mode)
- `tsh proxy db`

So no UX change to these.

In single port mode, using native clients to connect to Teleport Proxy directly with `tsh db env/config` will NOT work
through load balancers. Users can use one of the above methods instead.

#### 2 - Application Access UX

Application Access through Teleport Webapp is NOT affected by the issue of interest of this RFD.

On the `tsh` side, TCP, AWS, Azure, and GCP apps always use a local proxy so they will automatically perform the
connection upgrade when needed. No UX change to these.

However, HTTP apps MUST use `tsh proxy app`.

#### 3 - Kubernetes Access UX

Using native Kubernetes clients after `tsh kube login` will NOT work as the native clients cannot upgrade the
connection.

> `tsh kubectl` will automatically starts a local proxy and performs the connection upgrade when needed.

Reviewer: Do we plan to support standalone kubectl command ? 
I think at least  in case of running `kubectl` without kubeproxy we should consider printing description message that the `kubectl` requires running local `tsh proxy kube`

Author: Added explanation that `tsh kube login` should advertise `tsh xxxx`. I am considering in this case, we probably SHOULDN'T even update the default kube config like `~/.kube/config`.

Reviewer 2: We can support vanilla `kubectl` if the user exposes a new copy of the kubeconfig when using `tsh proxy kube .... -p ....`-

We have customers using `go`, `js`, and `python` clients so we must keep compatibility with them. `tsh proxy kube ...` must create a deep copy of the user's default kubeconfig and edit the required fields. Once it terminates, it cleans the kubeconfig so even if the user tries to use it, kubectl will fail because the file does not exist. The only thing the user must do is configure the `KUBECONFIG` env var in the new shell.

`tsh kubectl` can start the proxy, login, and create the new kubeconfig automatically for simplicity

A new `tsh proxy kube` command will be added to support other Kubernetes clients. The command will start a local proxy
and provide a config for Kubernetes clients to use:

```

> $ tsh proxy kube -p 8888

Reviewer: if the `tsh proxy kube` will also generate the `KUBECONFIG` it should also support kube cluster name parameter to allow a user to select a kube cluster where the request should be routed.

Author: added a paragraph on the flags.

---
authors: STeve Huang (xin.huang@goteleport.com)
state: draft
---

# RFD XXX - TLS Routing behind load balancers

## Required Approvals

* Engineering: @smallinsky && @r0mant

> * Product: @xin || @klizhentas

Reviewer 2: ```suggestion
* Product: @xinding33 || @klizhentas
```

* Security: @reedloden

## What

This RFD details how to support [TLS

> Routing](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md) behind load

Reviewer 3: ```suggestion
Routing](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md) behind layer 7 load
```

Is this more accurate?

> balancers.

Reviewer 4: ```suggestion
balancers for client protocols supported by Teleport.
```

## Why

> Allows simple single-port Teleport deployment when Telport needs to sit behind a load balancer.

Reviewer 4: ```suggestion
Allows simple single-port Teleport deployment when Teleport needs to sit behind an L7 load balancer.
```

## Details

### The challenge

Here is a typical setup that we want to enable by this RFD:
```
 ┌───┐
 │tsh│
 └┬──┘
  │
  │SSH,DB,Kube,App,etc.
  │
 ┌▼────────────┐
 │Load Balancer◄──┐
 └┬────────────┘  │Reversse tunnel,
  │               │Join,
  │Web port       │etc.
  │               │
 ┌▼───────┐     ┌─┴──────┐
 │Teleport│     │Teleport│
 │Proxy(s)│     │Agent(s)│
 └────────┘     └────────┘
```

The "load balancer" here could be a cloud-hosted service like AWS ALB, a Kubernetes ingress controller, or any other
similar network solutions.

TLS Routing already combines all proxy ports into a single port. However, most load balancers do NOT support passthrough
of custom ALPN, SNI, or mTLS, which are required by TLS Routing, when the load balancers are configured to terminate
TLS.

### The "connection upgrade"

Borrowed from the "WebSocket" design, the Teleport client can make a connection upgrade through a web API on the

> Teleport Proxy:

Reviewer 3: I would include @rosstimothy on this review to make sure this upgrade does not erase all the connection latency improvements he's made in the past year.

Author: Will try to add some benchmarks later this week to the performance section at the bottom. The extra latency is certainly noticeable in my previous POC.

Author: Some preliminary data collected based on `tc.GetClusterAlerts` from `tsh login`.
https://github.com/gravitational/teleport/blob/702a3ebd32390a4660a3631150c280647423d6ec/lib/client/api.go#L2069-L2085

Two conn upgrades are done, one for the proxy SSH client (`connectToProxy`), and one for the auth client for the GRPC call.

Without conn upgrade it took 450ms, and with conn upgrade it took 570ms. My ping to Proxy is about 30ms. So about a 27% increase in latency for this particular case. ~ 2 RTT per upgrade.

Sad news in my POC, 9 conn upgrades are done per `tsh login`.

Author: Updated the performance section a little bit. The original plan was to implement the basic feature first and then look at possible performance improvement separately, but love to hear your thoughts on that.

Load balancers usually terminate TLS with certificates signed by publicly trusted CAs. Since Teleport Proxy is no longer
the public entry point, it is no longer neccessary for Teleport Proxy to serve a publicly trusted cert (e.g. using
ACME). Thus Teleport Proxy should be welcome to use self-signed certs when sitting behind load balancers.

To accommodate that, instead of serving the **Web** certs, the ALPN server on the Proxy will serve the **Host** certs
when receiving connections from the upgrade flow. And instead of using the SystemCertPool, the Teleport client should
use the **Host** CA for verifying the Proxy server when dialing TLS Routing inside an upgrade connection. Note that this
modification does NOT apply to Teleport ALPN protocols that already do mTLS with other certs.

### When to make a connection upgrade

A new **optional** config parameter (e.g. `tsh_routing_upgrade_mode`) will be added in `proxy_service` section in the
config yaml. This new config has no effect on how Proxy server serve incoming traffic but it will provide a hint to all
the clients through the webapi ping:

```
{
  "auth": {
  ...
  },
  "proxy": {
  ...
    "tls_routing_enabled": true,

>     "tls_routing_upgrade_mode": "auto"

Reviewer 4: Could we potentially combine these into a single parameter? For example:

```
tls_routing_mode: (on|off|alb)
```

Reviewer: `alb` is quite narrow term. 
Also wonder if should detect if teleport proxy is behind L7 LB automaticity per tsh login call and make this option like "hard  enforcement" of TLS routing connection upgrade where we can still do the auto check  inside no latency crucial places like: `tsh login` call and allow to enforce Connection mode upgrade on all clients by this flag. 

This will allow a bit to simplify the UX teleport proxy configuration because it would be quite hard to describe all all possible situations when the `tls_routing_mode` should be set to `alb` mode.

Author: > tsh login call and allow to enforce Connection mode upgrade on all clients by this flag.
Ya, i was thinking about this too. we can have the test result in the profile yaml per `tsh login`. Let me drop the mode then, and we can just cache the result where it's needed.

Author: I just had an idea for improving performance for the test by merging the `/webapi/ping` and ALPN test in one call.

When client calls `/webapi/ping`, it will put these in ALPN `[]string{"teleport-webapi-ping", "http/1.1"}`.

The Proxy server will serve both ALPN values with the same response. 

The only difference is the `NegotiatedProtocol` would be `teleport-webapi-ping` if there is no LB, and it would be `http/1.1` or empty if LB is terminating the TLS.

What do you guys think?

Author: removed the mode for simplicity and added a section for performance on the detection test.

  },
  ...
}
```

> `auto` is the default mode where the client performances a test to decide if a connection upgrade is necessary. The test

Reviewer 4: ```suggestion
`auto` is the default mode where the client performs a test to decide if a connection upgrade is necessary. The test
```

 │                         │                 │                             │
 │                     ┌───▼─────┐        ┌──▼──────┐                      │
 │                     │Handshake│        │Handshake├───────────────┐      │
 │ ┌───────────────────┤ Success │        │ Failed  │               │      │
 │ │                   └───┬─────┘        └┬────────┘               │      │
 │ │                       │               │                        │      │
 │ │Negotiated             │No ALPN        │remote error: tls:      │other │
 │ │teleport-reversetunnel │Negotiated     │no application protocol │errors│
 │ │                       │               │                        │      │
 │ │                   ┌───▼────┐          │                  ┌─────▼──┐   │
 │ │                   │Upgrade ◄──────────┘                  │  Not   │   │
 └─┼───────────────────►Required│                             │Required◄───┘
   │                   └────────┘                             └────▲───┘
   │                                                               │
   └───────────────────────────────────────────────────────────────┘
```

> Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,

Reviewer 4: ```suggestion
Connection upgrade for TLS Routing should be required when the client and the server fail to negotiate the ALPN protocol,
```

hinting there is a load balancer in the middle that terminates TLS.

> However, this test may not be bullet proof. In particular, the test explicitly looks for `tls: no application protocol`

Reviewer 4: Is there a way to utilize TLS alert codes somehow to determine this? Looks like "no application protocol" alert has code 120, no idea if we can get this info on the client side though:

https://github.com/golang/go/blob/master/src/crypto/tls/alert.go#L50

Author: seems `alertNoApplicationProtocol` is only used internally on the TLS server side.

from the remote when handshakes fails, but it is possible that a load balancer implementation decides to use a different
text.

> In such cases, setting the mode to `on` can enforce the clients to always do the connection upgrade. Also, setting the

Reviewer 4: Will this setting also be a part of cluster networking config resource? If so, let's give an example, if not and we don't want to allow to change it dynamically, let's mention it too.

mode `on` or `off` improves performance by skipping the auto test. An environment variable (e.g.
`TELEPORT_TLS_ROUTING_UPGRADE_MODE=off`) can be set to overwrite the mode per client, just in case.

### Supporting all TLS Routing protocols

In general, in ALL existing places we are dialing the TLS Routing connection, common logic will be added to the
`ContextDialer` to automatically upgrade the connection when necessary.

However, not all connections are initiated by `tsh`, `teleport`, or a client usingthe API lib. In such cases, a [local
Teleport
proxy](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md#local-teleport-proxy)
is required to dial the TLS Routing connection with the upgrade.

Local proxy has already been implemented for Database Access and Application Access. Here are some details how it can be
implemented for Kubernetes Access.

#### Local proxy for Kubernetes Access

`tsh` will serve a local HTTPS proxy server with a localhost self-signed CA. An ephemeral (per `tsh` command) KUBECONFIG
will be provided to Kubernetes applications to connect to the local server. This KUBECONFIG points to the local proxy

> using the cluster `server` field, and the config will include static user credentials generated with the local

Reviewer 2: `kubeconfig` has a field to specify the `proxy` URL using `proxy-url`. 

If we know that the cluster is using the ALB routing mode, we can force that field. It will help `kubectl`  to print a better error message if the proxy is not running locally instead of trying to dial a cluster that he does not know the expected protocol. For other use cases such as creating an MFA session without being prompted to introduce the MFD, we can use the `HTTPS_PROXY` env var

This change should be applied to teleport clusters and not to other direct clusters the user might have configured in  his kubeconfig

Author: Any advantage to using `proxy-url` to point to the local proxy over using `server` to point to the local proxy directly, for teleport clusters?

Reviewer 2: Not really. 
When I initially thought about the issue I imagined it using the default `.kube/config` file and on that, it would be better to use the `proxy-url`. 

But I got your idea of copying the default kubeconfig, populating it with `tsh proxy kube <cluster1> <cluster2>` and storing it under `/tmp/kubeconfig-adhahdhashd.yaml`  with the server being the local server and the SNI identifying the cluster, Once `tsh proxy kube` starts you sign the cluster certificates and negotiate the MFA session.

In that scenario, we don't need the `proxy-url`.

self-signed CA.

```
 ┌────┐                 ┌─────────┐             ┌────────┐
 │kube│                 │tsh local│             │Teleport│
 │app │                 │  proxy  │             │ Proxy  │

>  └┬───┘                 └─┬───────┘             └──────┬─┘

Reviewer 4: Might make sense to include ALB in this diagram as well? The local proxy will handle connection upgrade?

  │                       │                            │
  │                       ├──┐                         │
  │                       │  │generate:                │
  │                       │  │local CA                 │
  │                       │  │local client credentials │

>   │                       │◄─┘local KUBECONFIG         │

Reviewer 4: Is generating CA and client creds going to happen on every request? This will likely decimate performance if so and might undo performance improvement efforts @tigrato is working on this quarter.

Reviewer 2: I thought of two options:

## Using certs to identify the cluster
 
I think we will require a certificate to be present because we must identify the Kubernetes cluster the user is trying to access if it's not the default context.
`tsh kube credentials` must generate a fake cert to identify the target cluster and the proxy can keep the real certificate (with MFA challenges)  in memory.

## Using SNI to send the cluster identification to the local proxy

Alternatively and probably the better approach, we can leverage the SNI when connecting to the local proxy to include the cluster name:
```
SNI: {kubeName}.{clusterName}.kube-local.my-teleport-cluster.com 
```
`{clusterName}` is the root or leaf cluster where the `kubeName` cluster exists.

This approach requires editing the kubeconfig and using a different `server` entry per Teleport cluster configured. This would be my choice and we can completely drop the `tsh exec` when we execute `kubectl` since the local proxy will handle the certs and extracts the details from the SNI requested.

Author: I love the SNI idea. The local CA and local static credentials can be generated once per tsh login, on demand. I believe each `tsh` local proxy session needs a separate kubeconfig anyway for the local proxy port.

it will be used internally for routing.

### User Experience

Once the connection upgrade support to all protocols are implemented, users can be recommended to upgrade their Teleport
clusters to single port mode if the cluster is currently serving separate ports.

All scenarios discussed in this section assumes Teleport sits behind a load balancer so connection upgrade for TLS
Routing is required.

#### 1 - Database Access UX

Common flows already use `tsh` local proxy (since v8) so they will automatically perform the connection upgrade:

- `tsh db connect` (single port mode)
- `tsh proxy db`

So no UX change to these.

In single port mode, using native clients to connect to Teleport Proxy directly with `tsh db env/config` will NOT work
through load balancers. Users can use one of the above methods instead.

#### 2 - Application Access UX

Application Access through Teleport Webapp is NOT affected by the issue of interest of this RFD.

On the `tsh` side, TCP, AWS, Azure, and GCP apps always use a local proxy so they will automatically perform the
connection upgrade when needed. No UX change to these.

> However, HTTP apps MUST use `tsh proxy app`.

Reviewer 4: Nit: Give an example with curl.

#### 3 - Kubernetes Access UX

When upgrade is required, using native Kubernetes clients after `tsh kube login` will NOT work as the native clients
cannot upgrade the connection. `tsh kube login` should advertise users to run `tsh kubectl` and `tsh proxy kube` instead
of `kubectl version`. `tsh kube credentials` should also error out and suggest using `tsh kubectl` and `tsh proxy kube`.

`tsh kubectl` will automatically starts a local proxy and performs the connection upgrade when needed.

A new `tsh proxy kube` command will be added to support other Kubernetes clients. The command will start a local proxy
and provide a config for Kubernetes clients to use:

```

> $ tsh proxy kube cluster1 cluster2 -p 8888

Reviewer 2: Will we force users to identify the clusters they want to access when they start the proxy or will we allow them to select some (and we automatically generate the certs) and if a new cluster is requested, the proxy can create the new certificate on the fly and do the MFA if required?

Author: Good question. I initially thought we can apply [DB MFA session](https://github.com/gravitational/teleport/blob/master/rfd/0090-db-mfa-sessions.md) solution to the local kube proxy (where it prompt MFA once when starting local proxy). But I neglected the fact MFA is required per kube cluster.

One option is, when per-session-MFA is enabled, `tsh kube credentials` will ask the corresponding local proxy if they have cached MFA cert for the requested kube cluster. If not, `tsh kube credentials` will prompt for MFA and forward it to the local proxy. The prompt should happen only once per kube cluster per local proxy session.

What do you think?

I am curious how well the MFA prompt from `tsh kube credentials` works today for various kube applications.

Started local Kubernetes server at https://localhost:8888.

Please use the following KUBECONFIG for your Kubernetes applications:
export KUBECONFIG=<path to kubeconfig>
```

`tsh proxy kube` should support all flags supported by `tsh kube login` like cluster names, `--as` etc. and can be run
independently WITHOUT running `tsh kube login` first. If `tsh kube login` has been run perviously, `tsh proxy kube`
should inherit the flags from the login if they are not specfied during `tsh proxy kube`.

A `--exec` flag will be provided to `tsh proxy kube` to execute a command backed by the local proxy:
```
$ tsh proxy kube --exec -- helm install my-chart
```
This avoids needing two terminal sessions to run a Kubernetes application.

In addition, we can provide tips to users to improve their UX by utilizing `alias`:
```
alias kubectl=`tsh kubectl`
alias helm=`tsh proxy kube --exec -- helm`
```

or [`tsh` aliases](https://github.com/gravitational/teleport/blob/master/rfd/0061-tsh-aliases.md):
```
aliases:
    "helm": "$TSH proxy kube --exec helm"
```

#### 4 - Server Access UX

> No UX change.

Reviewer 4: "No UX change" as in "tsh ssh will continue to just work even through ALB"? What about connecting with vanilla OpenSSH clients?

Author: 👍 i missed that. This is `tsh config` right? I think we have to always use `ProxyCommand`. I will look into this and update.

Author: The ssh config generated is already calling `tsh proxy ssh` for openssh client:
https://github.com/gravitational/teleport/blob/74591e4ad1a91764b2976a8b1066c3635d3e9dd7/lib/config/openssh/openssh.go#L48-L56

So I don't think we need to change anything here.

or maybe i misunderstood the question that you are referring to something else?

### Security

When upgrading the connection, Teleport client verifies the load balancer's TLS cert using SystemCertPool. And as
mentioned early, at the TLS Routing request, the Teleport client will be configured with a Teleport CA for verifying the
Proxy server.

There is no authentication at the connection upgrade request. Authentication is deferred to the TLS Routing request so
authentication remains the same as if there is no connection upgrade.

### Performance

The downside doing the connection upgrade is the performance penalty.

> The connection is already double encrypted with TLS Routing, and now it is tripple encrypted with connection upgrade.

Reviewer 4: ```suggestion
The connection is already double encrypted with TLS Routing, and now it is triple encrypted with connection upgrade.
```

Mordern processors should have no trouble doing the job, but concurrent TLS Routing requests with connection upgrades
may affect CPU usage.

The more noticeable impact is likely the latency incurred by the **extra roundtrips** by the connection upgrade. A quick
API call throub connection upgrade by `tsh` may take double the time as before, since the latency between `tsh` and
Teleport Proxy is usually more significant than the latency within the Teleport agents.

> Some ideas for improving performance includes implementing resuable upgraded connections and multiplexing concurrent TLS

Reviewer 4: ```suggestion
Some ideas for improving performance includes implementing reusable upgraded connections and multiplexing concurrent TLS
```

   │                               │                               │
```

This allows the client to tunnel the "original TLS Routing" call through a Layer 7 (HTTP) load balancer.

### Teleport Proxy with self-signed certs

Load balancers usually terminate TLS with certificates signed by publicly trusted CAs. Since Teleport Proxy is no longer
the public entry point, it is no longer necessary for Teleport Proxy to serve a publicly trusted cert (e.g. using ACME).
Thus Teleport Proxy should be welcome to use self-signed certs when sitting behind load balancers.

To accommodate that, instead of serving the **Web** certs, the ALPN server on the Proxy will serve the **Host** certs
when receiving connections from the upgrade flow. And instead of using the SystemCertPool, the Teleport client should
use the **Host** CA for verifying the Proxy server when dialling TLS Routing inside an upgrade connection. Note that
this modification does NOT apply to Teleport ALPN protocols that already do mTLS with other certs.

### When to make a connection upgrade

Teleport clients should first perform a test to decide if a connection upgrade is necessary. The test is done by sending a

> TLS handshake with a Teleport custom ALPN to the Proxy server:

Reviewer 5: Does this imply that every connection will need to wait for a completely separate TCP and TLS handshake first? That severely hampers the work done lately by @rosstimothy to reduce `tsh` latency as much as possible. Can we figure out some other way to achieve this?

Running normal, direct connections in parallel with the one that will attempt an upgrade might work, for instance - and it would be nice to also keep some short-ish local cache of addresses that are reachable directly or via connection upgrade, so we don't have to always try both.

Author: In the current implementation, we do cache the result in tsh profile so only doing this once per login.

The slow part is the upgrade itself, adding 2 round trips (TLS 1.3) per new connection.
